packages/core-contracts/contracts/utils/SafeCast.sol

```diff
@@ -0,0 +1,73 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/**
+ * @title TODO Utility that avoids silent overflow errors.
+ *
+ * Example silent overflow errors in Solidity:
+ * 1) If uint256 A = type(uint128).max + 1, down casting A to uint128 returns zero.
+ * 2) If int256 A = -1, casting A to uint256 returns type(uint256).max.
+ */
+library SafeCast {
+    error CastError(bytes32 fromType, bytes32 toType);
+
+    // Note: Overloading doesn't seem to work for similar types, i.e. int256 and int128, uint256 and uint128, etc, so explicitly naming the functions differently here.
+
+    function uint256toUint128(uint256 x) internal pure returns (uint128) {
+        if (x > type(uint128).max) {
+            revert("Failed cast uint256 to uint128");
```

Comment:
normally I wouldn't complain about stuff like this, but since this library is giong to be used at the deepest levels of most of our contracts, these strings unfortunaetly add a lot of extra weight, would be better to define a ton of errors instead, the ABIs can take it...

Reply:
Yeah, I actually tried to use custom errors, but was having trouble with them. Solidity started getting weird with all this low level stuff. Do you mind if we improve the errors in another PR?